DylibBinder/StreamWriterExtensions.cs

```diff
@@ -0,0 +1,887 @@
+			protocols.AddRange (p.FunctionsOfUnknownDestination.ToList ());
+			protocols.Sort ((type1, type2) => String.CompareOrdinal (type1.Signature.Name.ToString (), type2.Signature.Name.ToString ()));
+			var typedeclarationGenerics = new List<string> ();
+			var typedeclarationAssociatedTypes = new List<string> ();
+
+			var lastWrittenProtocolSignature = string.Empty;
+			foreach (var protocol in protocols) {
+				if (protocol.Signature.ToString () == lastWrittenProtocolSignature)
+					continue;
+				sw.WriteWithIndent ($"<func");
+				Indent ();
+				sw.WriteTypeValue ("name", StringBuilderHelper.EscapeCharacters (protocol.Signature.Name.ToString ()));
+				sw.WriteTypeValue ("operatorKind", protocol.Operator.ToString ());
+				sw.WriteTypeValue ("isStatic", CheckStaticProtocolMethod (protocol).ToString ());
+
+				var returnGenericChecker = sw.WriteReturnType (protocol.Signature, 0, protocol.Signature.GenericArguments.Count, protocol.Signature.GenericParameterCount);
+				var returnAssociateChecker = CheckParameterAssociateTypes (protocol.Signature.ReturnType, protocol.Signature, 0);
+				sw.WriteTypeValue ("hasThrows", protocol.Signature.CanThrow.ToString ());
+				sw.WriteTypeValue ("isPossiblyIncomplete", "True");
+
+				sw.WriteConstantAttributes ();
+				var parameterGenericChecker = sw.WriteParameters (protocol.Signature, 0);
+				var parameterAssociateChecker = CheckParameterAssociateTypes (protocol.Signature.Parameters, protocol.Signature, 0);
+
+				CheckTypeDeclarationGenerics (returnGenericChecker, parameterGenericChecker, null, ref typedeclarationGenerics, 0);
+				CheckTypeDeclarationAssociatedTypes (returnAssociateChecker, parameterAssociateChecker, ref typedeclarationAssociatedTypes);
+				Exdent ();
+				sw.WriteLineWithIndent ($"</func>");
+
+				lastWrittenProtocolSignature = protocol.Signature.ToString ();
+			}
+			return (typedeclarationGenerics, typedeclarationAssociatedTypes);
+		}
+
+		static List<string> CheckReturnAssociateTypes (SwiftReflector.SwiftType returnType)
+		{
+			if (returnType == null)
+				return null;
+
+			var retType = returnType as SwiftReflector.SwiftGenericArgReferenceType;
+			if (retType == null)
+				return null;
+
+			var associatedTypeList = new List<string> ();
+			foreach (var associatedType in retType.AssociatedTypePath) {
+				if (!associatedTypeList.Contains (associatedType)) {
+					associatedTypeList.Add (associatedType);
+				}
+			}
+			return associatedTypeList;
+		}
+
+		static List<string> CheckParameterAssociateTypes (object parameters, SwiftReflector.SwiftBaseFunctionType signature, int depth)
+		{
+			if (parameters == null)
+				return null;
+
+			if (depth > 0 && ((signature.GenericParameterCount == 0 && signature.GenericArguments.Count == 0) && !(signature.ToString ().Contains ("0,") && signature.ToString ().Contains ("1,")))) {
+				return null;
+			}
+
+			var associatedTypeList = new List<string> ();
+
+			switch (parameters.GetType ().Name) {
```

Comment:
Switching on type name is can miss subclasses. SwiftFunctionType, IIRC is subclassed. You probably want this structure instead:
```
if (parameters is SwiftGenericArgReferenceType genArg) {
    HandleAssociateGenericArgType (genArg, ref associateTypeList);
} else if (parameters is SwiftTupleType tuple) {
...
}
```

Reply:
Ah so much better!